# PYTHON: DATA MANAGEMENT TIPS OVER CONNECTIVITY

# PostgreSQL Platform

By: Hector Alvaro Rojas &nbsp;&nbsp;|&nbsp;&nbsp; Data Science, Visualizations and Applied Statistics &nbsp;&nbsp;|&nbsp;&nbsp; January 20, 2018<br>
    Url: [http://www.arqmain.net]   &nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;   GitHub: [https://github.com/arqmain]
    <hr>

## 1 How about getting the Connection?

I decide use "SQLAlchemy" to keep the standar of the examples done with "MySQL", SQLite" and "MS-SQL".  They accept "SQLAlchemy" to get "df.to_sql" when using Pandas.

In [3]:
# sqlalchemy connection using psycopg2
import psycopg2
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')


## 2 How about Querying the Database?

### 21 Selecting all (*)

In [4]:
import pandas as pd
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')

sql = "SELECT * FROM innova_first"
df = pd.read_sql_query(sql, engine)
print (df.shape)
df.head()

(3545, 13)


rol  identificador  region  ventas_ao_2005  expo_ao_2005  \
0  110467.0            2.0     1.0       7135775.0           0.0   
1  110468.0            2.0     1.0        367582.0           0.0   
2  110469.0            2.0     1.0       1650613.0           0.0   
3  110472.0            2.0     1.0      62272395.0    38644399.0   
4  110473.0            2.0     1.0       1191963.0           0.0   

   empleo_ao_2005  ventas_ao_2006  expo_ao_2006  empleo_ao_2006  producto1  \
0           472.0       6745463.0     2091093.0           411.0        2.0   
1             7.0        122981.0           0.0             7.0        1.0   
2            50.0       1769443.0           0.0            50.0        2.0   
3           209.0      45995579.0    24126202.0           277.0        1.0   
4            85.0       1090426.0           0.0            80.0        2.0   

   producto2  producto3  producto4  
0        2.0        2.0        2.0  
1        2.0        2.0        2.0  
2        2.0        2.0        2.0  
3        2.0        2.0        2.0  
4        2.0        2.0        2.0

### 22 Selecting fields

In [5]:
import pandas as pd
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')

sql = "SELECT rol, region, ventas_ao_2005, empleo_ao_2005 FROM innova_first"
df = pd.read_sql_query(sql, engine)
print (df.shape)
df.head()

(3545, 4)


rol  region  ventas_ao_2005  empleo_ao_2005
0  110467.0     1.0       7135775.0           472.0
1  110468.0     1.0        367582.0             7.0
2  110469.0     1.0       1650613.0            50.0
3  110472.0     1.0      62272395.0           209.0
4  110473.0     1.0       1191963.0            85.0

### 23 Selecting fields and conditions

In [6]:
import pandas as pd
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')

sql = "SELECT rol, region, ventas_ao_2005, empleo_ao_2005  FROM innova_first WHERE ventas_ao_2005 >=1500000 AND region >= 8"
df = pd.read_sql_query(sql, engine)
print (df.shape)
df.head()

(948, 4)


rol  region  ventas_ao_2005  empleo_ao_2005
0  210544.0    13.0      32952041.0           510.0
1  441584.0    13.0       8579325.0             4.0
2  810043.0     8.0      94220000.0           244.0
3  810238.0     8.0       6027533.0           223.0
4  810821.0     8.0      23870000.0           428.0

## 3 How about Creating a Database?

In [7]:
import pandas as pd
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')

# Select elements new table
sql = "SELECT rol, region, ventas_ao_2005, empleo_ao_2005 FROM innova_first"

# Made pd data frame ("paso")
paso = pd.read_sql_query(sql, engine)

# Create new table "first333" in "mybase1" database 
paso.to_sql(name='first333', con=engine, if_exists='replace',index=False)

# Verify if new table "first333" was created in "mybase1" database
sql = "SELECT * FROM first333"
df = pd.read_sql_query(sql, engine)
print (df.shape)
df.head()

(3545, 4)


rol  region  ventas_ao_2005  empleo_ao_2005
0  110467.0     1.0       7135775.0           472.0
1  110468.0     1.0        367582.0             7.0
2  110469.0     1.0       1650613.0            50.0
3  110472.0     1.0      62272395.0           209.0
4  110473.0     1.0       1191963.0            85.0

## 4 How about Quering by JOIN and Aggregations Summary?

### 41 Joining tables with JOIN

In [8]:
### Simple JOIN

import pandas as pd
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')

sql = "SELECT innova_first.rol, innova_first.region, tamano, ventas_ao_2005 from innova_first inner join innova_second on innova_first.rol = innova_second.rol"
df = pd.read_sql_query(sql, engine)
print (df.shape)
df.head()

(3539, 4)


rol  region  tamano  ventas_ao_2005
0  110467.0     1.0     2.0       7135775.0
1  110468.0     1.0     1.0        367582.0
2  110473.0     1.0     2.0       1191963.0
3  110474.0     1.0     2.0     125114709.0
4  110476.0     1.0     2.0      24413000.0

In [16]:
### JOIN with agregated summary

import pandas as pd
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')

sql = "SELECT innova_first.region, tamano, AVG(ventas_ao_2005) AS AVERAGE,  COUNT(ventas_ao_2005) as MUESTRA,  Stddev(ventas_ao_2005) AS DEstandar, MAX(ventas_ao_2005) AS MAXIMO, MIN(ventas_ao_2005) AS MINIMO from innova_first  inner join innova_second on innova_first.rol = innova_second.rol GROUP BY  innova_first.region, tamano  order BY innova_first.region, tamano"
df = pd.read_sql_query(sql, engine) 
print (df.shape) 
df

(26, 7)


region  tamano       average  muestra     destandar        maximo  \
0      1.0     1.0  1.514777e+07       15  4.115188e+07  1.620133e+08   
1      1.0     2.0  1.909793e+07      120  1.076017e+08  8.251621e+08   
2      2.0     1.0  1.535026e+07       19  5.093042e+07  2.237893e+08   
3      2.0     2.0  5.943262e+07      143  3.205218e+08  2.975540e+09   
4      3.0     1.0  1.313349e+07        9  2.564381e+07  6.177134e+07   
5      3.0     2.0  1.262486e+07      101  3.186912e+07  2.066203e+08   
6      4.0     1.0  8.675473e+06       13  1.773121e+07  4.883060e+07   
7      4.0     2.0  1.234347e+07      131  9.686322e+07  1.097700e+09   
8      5.0     1.0  1.110714e+07       31  1.586639e+07  6.559494e+07   
9      5.0     2.0  2.080806e+07      257  1.053828e+08  1.114396e+09   
10     6.0     1.0  2.255674e+07       18  3.800087e+07  1.531200e+08   
11     6.0     2.0  5.003657e+07      136  3.722050e+08  4.140368e+09   
12     7.0     1.0  6.073176e+06       28  1.295556e+07  6.204710e+07   
13     7.0     2.0  8.534253e+06      166  3.063806e+07  2.534379e+08   
14     8.0     1.0  1.131633e+07       63  2.858910e+07  1.989449e+08   
15     8.0     2.0  1.614574e+07      281  9.291718e+07  1.478137e+09   
16     9.0     1.0  7.324416e+06       27  1.473889e+07  6.328700e+07   
17     9.0     2.0  1.987242e+06      143  8.899381e+06  1.037251e+08   
18    10.0     1.0  5.477151e+07       28  1.848306e+08  9.800000e+08   
19    10.0     2.0  1.012062e+07      240  5.678005e+07  8.472237e+08   
20    11.0     1.0  1.038834e+07        8  2.055150e+07  5.691669e+07   
21    11.0     2.0  4.126802e+06       59  9.529513e+06  4.135230e+07   
22    12.0     1.0  9.556414e+05        8  1.036184e+06  2.943442e+06   
23    12.0     2.0  1.067896e+07       78  5.560789e+07  3.662239e+08   
24    13.0     1.0  4.943614e+07      294  3.558578e+08  5.962235e+09   
25    13.0     2.0  2.437065e+07     1123  2.074564e+08  5.532432e+09   

      minimo  
0   133011.0  
1        0.0  
2   200000.0  
3        0.0  
4    45000.0  
5        0.0  
6    52888.0  
7     3360.0  
8   107323.0  
9        0.0  
10  105706.0  
11    3000.0  
12   15000.0  
13       0.0  
14   21000.0  
15       0.0  
16    5000.0  
17       0.0  
18    1800.0  
19       0.0  
20  143332.0  
21       0.0  
22  142448.0  
23    7245.0  
24   28000.0  
25       0.0

### 42 Joining tables with WHERE

In [17]:
### Simple JOIN

import pandas as pd
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')

sql = "SELECT innova_first.rol, innova_first.region,  ventas_ao_2005 from innova_first, innova_second WHERE innova_first.rol = innova_second.rol"
df = pd.read_sql_query(sql, engine)
print (df.shape)
df.head()

(3539, 3)


rol  region  ventas_ao_2005
0  110467.0     1.0       7135775.0
1  110468.0     1.0        367582.0
2  110473.0     1.0       1191963.0
3  110474.0     1.0     125114709.0
4  110476.0     1.0      24413000.0

In [ ]:
conn.close()

In [19]:
### JOIN with agregated summary

import pandas as pd
from sqlalchemy import create_engine

# Connect to the database
engine = create_engine('postgresql+psycopg2://postgres:Alvaro@localhost/postgres')

sql = "SELECT innova_first.region, innova_second.tamano, AVG(ventas_ao_2005) AS AVERAGE, COUNT(ventas_ao_2005) as MUESTRA, Stddev(ventas_ao_2005) AS DEstandar, MAX(ventas_ao_2005) AS MAXIMO, MIN(ventas_ao_2005) AS MINIMO from innova_first, innova_second WHERE innova_first.rol = innova_second.rol GROUP BY   innova_first.region, innova_second.tamano  order BY innova_first.region, innova_second.tamano"
df = pd.read_sql_query(sql, engine)
print (df.shape)
df

(26, 7)


region  tamano       average  muestra     destandar        maximo  \
0      1.0     1.0  1.514777e+07       15  4.115188e+07  1.620133e+08   
1      1.0     2.0  1.909793e+07      120  1.076017e+08  8.251621e+08   
2      2.0     1.0  1.535026e+07       19  5.093042e+07  2.237893e+08   
3      2.0     2.0  5.943262e+07      143  3.205218e+08  2.975540e+09   
4      3.0     1.0  1.313349e+07        9  2.564381e+07  6.177134e+07   
5      3.0     2.0  1.262486e+07      101  3.186912e+07  2.066203e+08   
6      4.0     1.0  8.675473e+06       13  1.773121e+07  4.883060e+07   
7      4.0     2.0  1.234347e+07      131  9.686322e+07  1.097700e+09   
8      5.0     1.0  1.110714e+07       31  1.586639e+07  6.559494e+07   
9      5.0     2.0  2.080806e+07      257  1.053828e+08  1.114396e+09   
10     6.0     1.0  2.255674e+07       18  3.800087e+07  1.531200e+08   
11     6.0     2.0  5.003657e+07      136  3.722050e+08  4.140368e+09   
12     7.0     1.0  6.073176e+06       28  1.295556e+07  6.204710e+07   
13     7.0     2.0  8.534253e+06      166  3.063806e+07  2.534379e+08   
14     8.0     1.0  1.131633e+07       63  2.858910e+07  1.989449e+08   
15     8.0     2.0  1.614574e+07      281  9.291718e+07  1.478137e+09   
16     9.0     1.0  7.324416e+06       27  1.473889e+07  6.328700e+07   
17     9.0     2.0  1.987242e+06      143  8.899381e+06  1.037251e+08   
18    10.0     1.0  5.477151e+07       28  1.848306e+08  9.800000e+08   
19    10.0     2.0  1.012062e+07      240  5.678005e+07  8.472237e+08   
20    11.0     1.0  1.038834e+07        8  2.055150e+07  5.691669e+07   
21    11.0     2.0  4.126802e+06       59  9.529513e+06  4.135230e+07   
22    12.0     1.0  9.556414e+05        8  1.036184e+06  2.943442e+06   
23    12.0     2.0  1.067896e+07       78  5.560789e+07  3.662239e+08   
24    13.0     1.0  4.943614e+07      294  3.558578e+08  5.962235e+09   
25    13.0     2.0  2.437065e+07     1123  2.074564e+08  5.532432e+09   

      minimo  
0   133011.0  
1        0.0  
2   200000.0  
3        0.0  
4    45000.0  
5        0.0  
6    52888.0  
7     3360.0  
8   107323.0  
9        0.0  
10  105706.0  
11    3000.0  
12   15000.0  
13       0.0  
14   21000.0  
15       0.0  
16    5000.0  
17       0.0  
18    1800.0  
19       0.0  
20  143332.0  
21       0.0  
22  142448.0  
23    7245.0  
24   28000.0  
25       0.0

## References:
>* This is SQLAlchemy web site [SQLAlchemy 1.1 Documentation](http://docs.sqlalchemy.org/en/latest/core/engines.html) to get the necessary formal structure of the syntax from section <b>PostgreSQL</b>.<br>
>* From here [pandas.DataFrame.to_sql](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_sql.html) we get the parameters definition.
>* From here [PYTHON: PANDAS → MYSQL USING SQLALCHEMY. A.K.A SQLALCHEMY FOR PANDAS USERS WHO DON’T KNOW SQL (THE BRAVE AND THE FOOLHARDY)](https://robertdavidwest.com/2014/10/12/python-pandas-%E2%86%92-mysql-using-sqlalchemy-a-k-a-sqlalchemy-for-pandas-users-who-dont-know-sql-the-brave-and-the-foolhardy/) we get other material too.<br>
>* This is SQLAlchemy web site [SQLAlchemy 1.1 Documentation](http://docs.sqlalchemy.org/en/latest/core/engines.html) to get the necessary formal structure of the syntax.<br>
>* To get a summary of python aggregated functions see this link: [Pandas Essential Basic Functionality](http://pandas.pydata.org/pandas-docs/stable/basics.html) in the <b><u>Descriptive statistics</u></b> section.

<hr>
By: Hector Alvaro Rojas &nbsp;&nbsp;|&nbsp;&nbsp; Data Science, Visualizations and Applied Statistics &nbsp;&nbsp;|&nbsp;&nbsp; January 20, 2018<br>
    Url: [http://www.arqmain.net]   &nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;   GitHub: [https://github.com/arqmain]
    <hr>